common/repositories/guild_members_repository.h

```diff
@@ -0,0 +1,193 @@
+		Database& db,
+		uint32	char_id,
+		uint32	rank_id
+	) {
+		auto results = db.QueryDatabase(
+			fmt::format("UPDATE `{}` SET `rank`= '{}' WHERE `char_id` = '{}';",
+				TableName(),
+				rank_id,
+				char_id
+			)
+		);
+
+		return results.Success() ? results.RowsAffected() : 0;
+	}
+
+	static int ReplaceOne(
+		Database& db,
+		const GuildMembers& e
+	)
+	{
+		std::vector<std::string> v;
+
+		auto columns = Columns();
+
+		v.push_back(std::to_string(e.char_id));
+		v.push_back(std::to_string(e.guild_id));
+		v.push_back(std::to_string(e.rank));
+		v.push_back(std::to_string(e.tribute_enable));
+		v.push_back(std::to_string(e.total_tribute));
+		v.push_back(std::to_string(e.last_tribute));
+		v.push_back(std::to_string(e.banker));
+		v.push_back("'" + Strings::Escape(e.public_note) + "'");
+		v.push_back(std::to_string(e.alt));
+
+		auto results = db.QueryDatabase(
+			fmt::format(
+				"REPLACE INTO {} ({}) VALUES({})",
+				TableName(),
+				ColumnsRaw(),
+				Strings::Implode(", ", v)
+			)
+		);
+
+		return (results.Success() ? results.RowsAffected() : 0);
+	}
+
+	static int UpdateEnabled(Database& db, uint32 guild_id, uint32 char_id, uint32 enabled)
+	{
+		auto results = db.QueryDatabase(
+			fmt::format(
+				"UPDATE {} SET `tribute_enable` = '{}' WHERE `char_id` = '{}' AND `guild_id` = '{}'",
+				TableName(),
+				enabled,
+				char_id,
+				guild_id
+			)
+		);
+
+		return (results.Success() ? results.RowsAffected() : 0);
+	}
+
+	static std::vector<GuildMembersWithTributeOn> GetMembersWithTributeOn(Database& db, uint32 guild_id)
+	{
+		std::vector<GuildMembersWithTributeOn> all_entries;
+
+		auto results = db.QueryDatabase(
+			fmt::format("SELECT gm.char_id, gm.guild_id, gm.tribute_enable, cd.`name`, cd.`level`, cd.deleted_at "
+				"FROM guild_members gm "
+				"JOIN character_data cd ON cd.id = gm.char_id "
+				"WHERE ISNULL(cd.deleted_at) AND gm.tribute_enable = 1 AND gm.guild_id = '{}';",
+				guild_id
+				)
+		);
+
+		all_entries.reserve(results.RowCount());
+
+		for (auto row = results.begin(); row != results.end(); ++row) {
+			GuildMembersWithTributeOn e{};
+
+			e.char_id		 = static_cast<int32_t>(atoi(row[0]));
+			e.guild_id		 = static_cast<uint32_t>(strtoul(row[1], nullptr, 10));
+			e.tribute_enable = static_cast<uint8_t>(strtoul(row[2], nullptr, 10));
+			e.char_name		 = row[3] ? row[3] : "";
+			e.char_level	 = static_cast<uint32_t>(strtoul(row[4], nullptr, 10));
+			
+			all_entries.push_back(e);
+		}
+
+		return all_entries;
+	}
+
+	static int UpdateFavor(Database& db, uint32 guild_id, uint32 char_id, uint32 favor)
+	{
+		auto results = db.QueryDatabase(
+			fmt::format(
+				"UPDATE {} SET `total_tribute` = '{}', `last_tribute` = '{}' WHERE `char_id` = '{}' AND `guild_id` = '{}'",
+				TableName(),
+				favor,
+				time(nullptr),
+				char_id,
+				guild_id
+			)
+		);
+
+		return (results.Success() ? results.RowsAffected() : 0);
+	}
+
+	static int UpdateOnline(Database& db, uint32 char_id, bool status)
```

Comment:
Your base repository was missing the `online` field so I regenerated it for you and pushed it up

Instead of doing raw queries in custom repository methods, you can lean into the core functionality of repositories here.

```cpp
static int UpdateOnline(Database& db, uint32 char_id, bool status)
{
	const auto guild_members = GetWhere(db, fmt::format("char_id = '{}'", char_id));
	if (guild_members.empty()) {
		return 0;
	}

	auto m = guild_members[0];

	m.online = status ? 1 : 0;

	return UpdateOne(db, m);
}
```

Needs to be tested of course but should work. You are querying and use an active record pattern to manipulate the struct and then pass the data model back to an update method and the repository method takes care of doing the update for you. You're doing a lot of the work the base repositories are already doing

Reply:
Same thing goes for other methods - you can re-use the same pattern and lean into the core repository methods

Reply:
Refactored all to better use the repos.  Thanks for the advice.